### PR TITLE
MEN-1187: Spontanaeous-reboot hardening of the client

### PR DESCRIPTION
**Pull request by oleorhagen**
Changelog: Title

Now the client, if losing power, will fall back to the original partition,
and run the Artifact-failure scripts, before going back to idle

Signed-off-by: Ole Petter <ole.orhagen@cfengine.com>

### CODE REVIEW

**Comment by oleorhagen** [comment]
Welcome back! @pasinskim

File: mender.go [hunk 1]
```diff
@@ -672,6 +672,12 @@ func (m *mender) TransitionState(to State, ctx *StateContext) (State, bool) {
 
 		m.SetNextState(to)
 
+		// Special case cleanup of stored state-data
```
Comment (pasinskim) [inline mender.go]: Can you add some comment why do we need this.

File: state.go [hunk 2]
```diff
@@ -356,24 +356,52 @@ func (i *InitState) Handle(ctx *StateContext, c Controller) (State, bool) {
...
 
 		return NewUpdateErrorState(me, sd.UpdateInfo), false
 	}
 }
 
+func committedPartition(c Controller) bool {
+
+	ua, err := c.HasUpgrade()
+	log.Infof("in committedPartition: ua: %t, err: %v", ua, err)
+	if err != nil {
```
Comment (pasinskim) [inline state.go]: Maybe we can print some error message here as this is pretty important and will impact state transition.

File: state.go [hunk 3]
```diff
@@ -356,24 +356,52 @@ func (i *InitState) Handle(ctx *StateContext, c Controller) (State, bool) {
 
 	log.Infof("handling loaded state: %s", sd.Name)
 
-	// chack last known state
-	switch sd.Name {
-	// update process was finished; check what is the status of update
-	case MenderStateReboot:
+	if !committedPartition(c) {
+		// only valid entrypoint into the uncommitted partition is reboot_leave
```
Comment (pasinskim) [inline state.go]: Is `commit_enter` also a possibility here? What I am thinking is we've entered `commit_enter` state and then the reboot happened. Will we end up here as well?

File: state.go [hunk 4]
```diff
@@ -356,24 +356,52 @@ func (i *InitState) Handle(ctx *StateContext, c Controller) (State, bool) {
...
+
+	// check last known state
+	switch sd.Name {
 
 	case MenderStateRollbackReboot:
 		return NewAfterRollbackRebootState(sd.UpdateInfo), false
 
-	// this should not happen
+	// Rerun commit-leave
+	case MenderStateUpdateCommit:
```
Comment (pasinskim) [inline state.go]: Do you want to force the update_commit -> init transition here? Why can't we simply skip setting the update_commit state?

File: state.go [hunk 5]
```diff
@@ -922,16 +959,19 @@ func (usr *UpdateStatusReportState) Handle(ctx *StateContext, c Controller) (Sta
...
-		Name:         usr.Id(),
-		UpdateInfo:   usr.Update(),
-		UpdateStatus: usr.status,
-	}); err != nil {
-		log.Errorf("failed to store state data in update status report state: %v",
-			err)
-		return NewReportErrorState(usr.Update(), usr.status), false
+	// Do not store this if artifact-commit scripts are run when leaving the state
+	// as then the scripts will not be rerun
+	if usr.Transition() != ToArtifactCommit {
```
Comment (pasinskim) [inline state.go]: So why we are storing the state now only if we are not entering here from commit state? 
Next, the fact that we need to use `Transition()` here means we need to really think about refactoring this whole state machine concept as it is becoming quite messy.

**Comment by pasinskim** [comment]
@oleorhagen maybe we can rerun integration tests as some state scripts tests are failing here. Or maybe this is expected without some modifications in the tests itself.

**Pull request merged**